sdc-plugin/clocks.cc

```diff
@@ -19,220 +19,111 @@
 }
 
-void Clocks::Propagate(BufferPropagation* pass) {
-#ifdef SDC_DEBUG
-    log("Start buffer clock propagation\n");
-    log("IBUF pass\n");
-#endif
-    std::vector<Clock> clocks(clocks_);
-    for (auto clock : clocks) {
-#ifdef SDC_DEBUG
-	log("Processing clock %s\n", clock.Name().c_str());
-#endif
-	PropagateThroughBuffer(pass, clock, IBuf());
-    }
-#ifdef SDC_DEBUG
-    log("BUFG pass\n");
-#endif
-    clocks = clocks_;
-    for (auto clock : clocks) {
-#ifdef SDC_DEBUG
-	log("Processing clock %s\n", clock.Name().c_str());
-#endif
-	PropagateThroughBuffer(pass, clock, Bufg());
-    }
-#ifdef SDC_DEBUG
-    log("Finish buffer clock propagation\n\n");
-#endif
-}
-
-void Clocks::Propagate(ClockDividerPropagation* pass) {
-#ifdef SDC_DEBUG
-    log("Start clock divider clock propagation\n");
-#endif
-    for (auto clock : clocks_) {
-#ifdef SDC_DEBUG
-	log("Processing clock %s\n", clock.Name().c_str());
-#endif
-	auto pll_clocks =
-	    pass->FindSinkClocksForCellType(clock, "PLLE2_ADV");
-	for (auto pll_clock : pll_clocks) {
-#ifdef SDC_DEBUG
-	    log("PLL clock: %s\n", pll_clock.Name().c_str());
-#endif
-	    AddClock(pll_clock);
-	    PropagateThroughBuffer(pass, pll_clock, Bufg());
+std::pair<float, float> Clock::Waveform(RTLIL::Wire* clock_wire) {
+    if (!clock_wire->has_attribute(RTLIL::escape_id("WAVEFORM"))) {
+	float period(Period(clock_wire));
+	if (!period) {
+	    log_cmd_error(
+	        "Neither PERIOD nor WAVEFORM has been specified for wire %s\n",
+	        WireName(clock_wire).c_str());
+	    return std::make_pair(0, 0);
 	}
+	float falling_edge = period / 2;
+	log_warning(
+	    "Waveform has not been specified\n Default value {0 %f} will be "
+	    "used\n",
+	    falling_edge);
+	return std::make_pair(0, falling_edge);
     }
-#ifdef SDC_DEBUG
-    log("Finish clock divider clock propagation\n\n");
-#endif
-}
-
-void Clocks::PropagateThroughBuffer(Propagation* pass, Clock& clock,
-                                    Buffer buffer) {
-    auto clock_wires = clock.GetClockWires();
-    for (auto clock_wire : clock_wires) {
-#ifdef SDC_DEBUG
-	log("Clock wire %s\n", RTLIL::unescape_id(clock_wire->name).c_str());
-#endif
-	auto buf_wires = pass->FindSinkWiresForCellType(clock_wire, buffer.name,
-	                                                buffer.output);
-	int path_delay(0);
-	for (auto wire : buf_wires) {
-#ifdef SDC_DEBUG
-	    log("%s wire: %s\n", buffer.name.c_str(),
-	        RTLIL::unescape_id(wire->name).c_str());
-#endif
-	    path_delay += buffer.delay;
-	    AddClock(RTLIL::unescape_id(wire->name), wire, clock.Period(),
-	             clock.RisingEdge() + path_delay,
-	             clock.FallingEdge() + path_delay);
-	}
+    float rising_edge(0);
+    float falling_edge(0);
+    std::string waveform(
+        clock_wire->get_string_attribute(RTLIL::escape_id("WAVEFORM")));
+    if (std::sscanf(waveform.c_str(), "%f %f", &rising_edge, &falling_edge) != 2) {
+	log_cmd_error("Incorrect WAVEFORM format\n");
```

Comment:
Please print the invalid WAVEFORM too?

Reply:
More specifically, error messages should:
 - Say what is wrong
 - Include enough context to determine why the input was wrong
 - Provide enough context to determine what is right

In this case:
 - The WAVEFORM attribute was wrong (good)
 - However which wire had the bad attribute is not given (bad)
 - The bad attribute is not given (bad)
 - What a good WAVEFORM looks like is not given (bad)

Please improve the error message.

Reply:
Done.